internal/notifier/github.go

```diff
@@ -40,17 +37,24 @@ type GitHub struct {
 	Client       *github.Client
 }
 
-func NewGitHub(commitStatus string, addr string, token string, certPool *x509.CertPool) (*GitHub, error) {
+func NewGitHub(commitStatus string, addr string, token string, certPool *x509.CertPool, proxyURL string, providerName string, providerNamespace string, secretData map[string][]byte, tokenCache *pkgcache.TokenCache) (*GitHub, error) {
 	if len(token) == 0 {
-		return nil, errors.New("github token cannot be empty")
-	}
+		githubOpts, err := getGitHubAppOptions(providerName, providerNamespace, proxyURL, secretData, tokenCache)
+		if err != nil {
+			return nil, err
+		}
 
-	host, id, err := parseGitAddress(addr)
-	if err != nil {
-		return nil, err
+		if len(githubOpts) == 0 {
+			return nil, errors.New("github token or github app details must be specified")
+		}
+
+		token, err = getGitHubAppToken(githubOpts)
+		if err != nil {
+			return nil, err
+		}
 	}
```

Comment:
It seems to me that this entire `if len(token) == 0 {` block is exactly the same on both providers, so it looks to me that this entire block can be abstracted into a single function merging `getGitHubAppOptions()`, `getGitHubAppToken()` and everything in the middle. Let's call it `getGitHubToken(token, providerName, providerNamespace, proxyURL, secretData, tokenCache)`.

Even better, I can see even more common parts. To me it looks like the constructors of these two providers can almost entirely be merged into a single function `newGitHubClient(...)`. Then in the `github` provider we check if the `commitStatus` is empty right before creating the client.